Add IPv4 and IPv6 multicast implementation for UDPSocket

kegilbert opened the pull request:
Continuation of initial pull request that only added IPv4 support. Extensive rework largely took place in lwip_stack.c to add IPv4/IPv6 logic.

Tested by enabling IPv6 in mbed_lib and mbed_app.json using the following test script on two K64Fs hooked up to a common router, one acting as a broadcast node and the other as a receiver.

```c++
#include "mbed.h"
#include "EthernetInterface.h"

const char* MCAST_GRP = "FF02::1";
//const char* MCAST_GRP = "229.0.0.1";
const int MCAST_PORT = 5007;

#ifdef SEND
int main() {
    DigitalOut led(LED2);
    EthernetInterface eth;
    eth.connect();
    
    UDPSocket sock;
    sock.open(&eth);
    
    SocketAddress multicast_group;
    multicast_group.set_ip_address(MCAST_GRP);
    multicast_group.set_port(MCAST_PORT);
    
    char out_buffer[] = "very important data";
    while (true) {
        printf("Multicast to group: %s\r\n", MCAST_GRP);
        sock.sendto(multicast_group, out_buffer, sizeof(out_buffer));
        led = !led;
        Thread::wait(1000);
    }
}
#else
DigitalOut led(LED3);

void beat(void) {
    led = !led;
}

int main() {
    printf("Running recv...\r\n");
    Ticker heart;
    heart.attach(&beat, 0.25f);
    EthernetInterface eth;
    eth.connect();
    
    UDPSocket server;
    server.open(&eth);
    server.bind(MCAST_PORT);
    int ret = server.join_multicast_group(MCAST_GRP);
    if (ret != 0) {
        printf("Error joining the multicast group, error: %d\r\n", ret);
        while (true) {}
    }
   
    SocketAddress client;
    char buffer[256];
    while (true) {
        printf("\r\nWait for packet...\r\n");
        int n = server.recvfrom(&client, buffer, sizeof(buffer));
        printf("Packet from \"%s\": %s\r\n", client.get_ip_address(), buffer);
    }
}
#endif
```

![ipv4-multicast-1](https://user-images.githubusercontent.com/1060940/30835360-812a48a4-a21d-11e7-8b18-0421ba56cf63.PNG)
![ipv6-multicast-1](https://user-images.githubusercontent.com/1060940/30835359-812a46d8-a21d-11e7-8b96-c8ed20062b0d.PNG)

@kjbracey-arm Thoughts on this implementation?

@theotherjimmy @geky

tommikas commented:
Failing to compile with IAR:
````
23:49:50 [K64F_6LOWPAN IAR] [Warning] lwip_stack.c@298,18: [Pe111]: statement is unreachable
23:49:50 [K64F_6LOWPAN IAR] [Warning] lwip_stack.c@663,18: [Pe174]: expression has no effect
23:49:50 [K64F_6LOWPAN IAR] [Warning] lwip_stack.c@691,37: [Pe174]: expression has no effect
23:49:50 [K64F_6LOWPAN IAR] [Error] lwip_stack.c@1134,25: [Pe144]: a value of type "void const *" cannot be used to initialize an entity of type "nsapi_ip_mreq_t *"
23:49:50 [K64F_6LOWPAN IAR] [Warning] lwip_stack.c@1158,0: [Pe069]: integer conversion resulted in truncation
```

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1125,53 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                return NSAPI_ERROR_UNSUPPORTED;
+            }
+            err_t igmp_err;
+            nsapi_ip_mreq_t *imr = optval;
```
Missing const on imr.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1125,53 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+            }
+            #endif
+
+            return igmp_err;
```
Need to do error conversion from lwip to NSAPI here. I guess you could then initialise igmp_err to ERR_VAL to get NSAPI_ERROR_PARAMETER - no err_t currently maps to NSAPI_ERROR_UNSUPPORTED.

Seems reasonable - I kind of like distinguishing "I don't know this option" from "I don't like what you specified for this option". I'd argue all existing size checks returning UNSUPOPRTED should be PARAMETER too.

kjbracey-arm commented on features/netsocket/UDPSocket.h:
```diff
@@ -56,6 +56,21 @@ class UDPSocket : public Socket {
+     * @param address   Multicast group IP address
+     * @return          Negative error code on failure
+     */
+    int join_multicast_group(const SocketAddress &address);
```
Wondering if this should be in Socket. This is falling into a trap much of the rest of the class does in that stuff ends up in UDPSocket only because it's "not TCP". Other types of socket like raw or ICMP will want this API, and it's calling an IP-level socket option.

(And Socket is currently "IP socket", rather than generic socket)

kegilbert replied:
Moved it over to Socket. Added an overload in TCPSocket to return an error for TCPSocket objects, let me know if you'd rather remove that altogether or had a better implementation.

kjbracey-arm replied:
I don't think it is particularly important to block it on a TCP socket - it's weird but harmless. On some stacks it does work - it does make you join the group regardless. The TCP socket wouldn't process those packets itself, as it never processes multicasts, but other sockets and pings would start responding.

But if you do want to block it, it logically needs to be in setsockopt, looking at `s->conn->type` like the TCP-specific options do.

kjbracey-arm commented:
That FF02::1 test isn't very good - that's "all hosts", so any IPv6 node should already be a member of it without you adding it.

May be worth checking you're seeing the correct IGMP/MLD messages being sent.

kegilbert commented:
Updated the branch to fix the IAR issue and move the implementation to the Socket class. Also successfully retested on IPv6 with FF05::208 as an unassigned address.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1125,53 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                return NSAPI_ERROR_PARAMETER;
+            }
+            err_t igmp_err;
+            const nsapi_ip_mreq_t *imr = (const nsapi_ip_mreq_t*) optval;
```
Unnecessary C-style cast is unhelpful - would have hidden the const-correctness error previously reported.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1125,53 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                ip_addr_set_any(IP_IS_V6(&if_addr), &if_addr);
+            }
+
+            igmp_err = NSAPI_ERROR_UNSUPPORTED;
```
This needs to be an lwIP error value, as you're going to convert it.

kegilbert commented:
Addressed @kjbracey-arm's comments and added some basic tracking of joined multicast addresses in the Socket class. On Socket close, all joined multicast addresses are left before closing to create a defined behavior.

kjbracey-arm commented on features/netsocket/Socket.cpp:
```diff
@@ -83,11 +94,39 @@ int Socket::modify_multicast_group(const SocketAddress &address, nsapi_socket_op
 
 int Socket::join_multicast_group(const SocketAddress &address)
 {
+    if(_multicast_addr_count == SOCKET_MAX_MULTICAST_MEMBERSHIPS) {
```
This is the wrong layer - it's in the "friendly veneer" rather than the core socket option. It means direct use of the socket options won't track.

kjbracey-arm commented on features/netsocket/Socket.cpp:
```diff
@@ -83,11 +94,39 @@ int Socket::modify_multicast_group(const SocketAddress &address, nsapi_socket_op
+    while((_multicast_addr_registry & (0x0001 << index))) { index++; }
+
+    // Create copy of socketaddress to prevent dead pointers when reading back during leave group
+    _socket_multicast_register[index] = SocketAddress(address);
```
It would be a copy-assignment anyway with just `= address`, wouldn't it?

Regardless, I think you would need to be storing the actual multicast request structures, not just the socket address, as you want to track memberships per-interface.

You might want to check for existing matches - add group normally refuses if you've already joined on the same socket. And cancel the entry if the underlying join fails?

kjbracey-arm commented on features/netsocket/Socket.h:
```diff
@@ -225,6 +228,9 @@ class Socket {
     rtos::Mutex _lock;
+    uint32_t _multicast_addr_count;
+    uint32_t _multicast_addr_registry;
+    SocketAddress  _socket_multicast_register[SOCKET_MAX_MULTICAST_MEMBERSHIPS];
```
An array feels a little memory inefficient, especially if it was actually the multicast request with both group and interface address. Most sockets won't be doing any joins at all. 

Maybe have these fields in a malloced sublock you allocate on first join? Although I know some disfavour dynamic allocation.

kegilbert replied:
Agree, I wasn't sure what the stance on dynamic memory allocation was compared to static allocation in mbed generally so went with this option. Completely open to changing it.

kegilbert commented:
Moved the socket membership logging to the lwip_stack layer. Switched to dynamic memory allocation on first multicast address add to avoid extra memory on every Socket object that does not use multicast.
@kjbracey-arm

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -81,9 +97,24 @@ static struct lwip_socket *mbed_lwip_arena_alloc(void)
+            .imr_multiaddr = s->multicast_memberships[index].imr_multiaddr,
+            .imr_interface = s->multicast_memberships[index].imr_interface
+        };
+        mbed_lwip_setsockopt(0, s, 0, NSAPI_DROP_MEMBERSHIP, &address, sizeof(address));
```
You could just pass `&s->multicast_memberships[index]`, no need for the temporary.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1154,106 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                if(igmp_err == ERR_OK) {
+                    if(!s->multicast_memberships) {
+                        // First multicast join on this socket, allocate space for membership tracking
+                        s->multicast_memberships = malloc(sizeof(nsapi_ip_mreq_t) * LWIP_SOCKET_MAX_MEMBERSHIPS);
```
Need to check for memory allocation failure here.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1154,106 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                    next_free_multicast_member(s->multicast_memberships_registery, index);
+
+                    s->multicast_memberships[index].imr_multiaddr = imr->imr_multiaddr;
+                    s->multicast_memberships[index].imr_interface = imr->imr_interface;
```
`s->multicast_memberships[index] = *imr;` would do.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1154,106 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                        next_registered_multicast_member(s->multicast_memberships_registery, index);
+
+                        bool address_matched = true;
+                        for(int i = 0; i < NSAPI_IP_BYTES; i++) {
```
I'd be a bit less literal about this; rather than a manual loop, address_matched is just
```
memcmp(&s->multicast_memberships[index].imr_multiaddr, &imr->imr_multiaddr, sizeof(nsapi_addr_t)) == 0 &&
memcmp(&s->multicast_memberships[index].imr_interface, &imr->imr_interface, sizeof(nsapi_addr_t)) == 0
```
You could just memcmp the whole structure, if you assumed no padding, which I think is probably fine.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1154,106 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                        // First multicast join on this socket, allocate space for membership tracking
+                        s->multicast_memberships = malloc(sizeof(nsapi_ip_mreq_t) * LWIP_SOCKET_MAX_MEMBERSHIPS);
+                    } else if(s->multicast_memberships_count == LWIP_SOCKET_MAX_MEMBERSHIPS) {
+                        return ERR_BUF;
```
If failing here, you've already joined, and the reference counting is borked. Maybe you should check for space first and fill it in.

Then clear the bit if the actual call to lwIP fails.

kegilbert commented:
Reordered the registry updates and added in the additional allocation checks along with some minor cleanup.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1169,99 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                    return NSAPI_ERROR_NO_ADDRESS;
+                }
+
+                clear_multicast_member_registry_bit(s->multicast_memberships_registery, find_multicast_member(s, imr));
```
`member_pair_index` - don't need to find again.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1169,99 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+            }
+
+            igmp_err = ERR_USE; // Maps to NSAPI_ERROR_UNSUPPORTED
+            int32_t member_pair_index = find_multicast_member(s, imr);
```
This could be the same variable as `index` - you don't need to track both.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1169,99 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                    // First multicast join on this socket, allocate space for membership tracking
+                    s->multicast_memberships = malloc(sizeof(nsapi_ip_mreq_t) * LWIP_SOCKET_MAX_MEMBERSHIPS);
+                    if(!s->multicast_memberships) {
+                        return ERR_BUF;
```
You're mixing error codes types again - you should be returning NSAPI errors

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1123,6 +1169,99 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+            igmp_err = ERR_USE; // Maps to NSAPI_ERROR_UNSUPPORTED
+            int32_t member_pair_index = find_multicast_member(s, imr);
+
+            if(optname == NSAPI_ADD_MEMBERSHIP) {
```
Style - `if (`. Applies in a lot of the most recent change

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -56,7 +67,12 @@ static struct lwip_socket {
+    // Track multicast addresses subscribed to by this socket
+    nsapi_ip_mreq_t *multicast_memberships;
+    uint32_t         multicast_memberships_count;
+    uint32_t         multicast_memberships_registery;
```
Spelling - `registry`

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -56,7 +67,12 @@ static struct lwip_socket {
+    nsapi_ip_mreq_t *multicast_memberships;
+    uint32_t         multicast_memberships_count;
+    uint32_t         multicast_memberships_registery;
+} lwip_arena[MEMP_NUM_NETCONN] = {0};
```
`={0}` is meaningless. Static data is always default-initialised to zero.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -81,9 +97,21 @@ static struct lwip_socket *mbed_lwip_arena_alloc(void)
 {
     s->in_use = false;
+
+    while(s->multicast_memberships_count > 0) {
```
style - `while (`

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -1080,6 +1108,24 @@ static nsapi_size_or_error_t mbed_lwip_socket_recvfrom(nsapi_stack_t *stack, nsa
     return recv;
 }
 
+int32_t find_multicast_member(struct lwip_socket *s, const nsapi_ip_mreq_t *imr) {
```
Should be static. `s` could be const.

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -46,6 +48,15 @@ static nsapi_error_t mbed_lwip_err_remap(err_t err);
+    #define LWIP_SOCKET_MAX_MEMBERSHIPS 4
+#endif
+
+#define next_registered_multicast_member(registry, index)       while(!(registry & (0x0001 << index))) { index++; }
```
Magic macros seem a bit tricksy, and not gaining much here.

Maybe avoid confusion by making them static inline functions that return an updated index, rather than modifying index by value?

We assume C99 or later, so `static inline` is portable (and works the same as in C++03).

kjbracey-arm commented:
As it's nearly there I'd like more review from @geky, @c1728p9, @SeppoTakalo and @mikaleppanen.

mikaleppanen commented:
Looks good to me.

kegilbert commented:
Updated style and moved from macros to static inline functions

kegilbert commented:
/morph test

theotherjimmy commented:
@kegilbert That was not a morph test. :D

theotherjimmy commented:
/test mbed-os

kjbracey-arm commented on features/netsocket/Socket.cpp:
```diff
@@ -70,6 +70,28 @@ nsapi_error_t Socket::close()
+    mreq.imr_multiaddr = address.get_addr();
+    mreq.imr_interface = nsapi_addr_t();   // Default address, NSAPI_UNSPEC
+
+    return this->setsockopt(0, socketopt, &mreq, sizeof(mreq));
```
First parameter here should be `NSAPI_SOCKET`

kjbracey-arm commented on features/FEATURE_LWIP/lwip-interface/lwip_stack.c:
```diff
@@ -84,6 +116,18 @@ static struct lwip_socket *mbed_lwip_arena_alloc(void)
+        static uint32_t index = 0;
+        index = next_registered_multicast_member(s, index);
+
+        mbed_lwip_setsockopt(0, s, 0, NSAPI_DROP_MEMBERSHIP, &s->multicast_memberships[index],
```
Third parameter should technically be `NSAPI_SOCKET` - at some point `mbed_lwip_setsockopt` should enforce that.

First parameter should be written as `NULL`.

theotherjimmy commented:
@AnotherButler Another Doxygen update here.

kegilbert commented:
Added multithreading protection to lwip join/leave group calls.

kegilbert commented:
This still has the Needs Review tag, should I add someone else to the review list?

geky commented:
I think this is good to go then, @theotherjimmy, @0xc0170, should this be "needs: preceding PR"?

Unfortunatly this will have to wait until we get the network tests working again: https://github.com/ARMmbed/mbed-os/pull/5335

0xc0170 commented:
> I think this is good to go then, @theotherjimmy, @0xc0170, should this be "needs: preceding PR"?

I am not certain now, @kegilbert tests are green, so all should be good now?

geky commented:
Although this needs #5335 before all tests are ran

0xc0170 commented:
>Although this needs #5335 before all tests are ran

@kegilbert Can you help @yennster with the 5335 as this depends on it? Or anything else can be done for this PR to make it ready for intergration?

kegilbert commented:
I can take a look at it when I get back next week from GGC, don't have access to a Nucleo board at the moment.

kegilbert commented:
https://github.com/ARMmbed/mbed-os/pull/5335 was merged, rekicking off CI

kegilbert commented:
/morph build

kegilbert commented:
@0xc0170 @geky Tests passed with the new network test patches.

geky commented:
LGTM
